src/MerkleGen.sol

```diff
@@ -14,12 +14,109 @@ library MerkleGen {
+        for (uint256 i = 0; i < layers.length - 1; i++) {
+            // Exclude the last layer because it's the root
+            layer = layers[i];
+            (indices, subproof, source_flags) = proveSingleLayer(layer, indices);
+            proof_hashes = proof_hashes.extend(subproof);
+            proof_source_flags = proof_source_flags.extend(source_flags);
+        }
+
+        // Get leaves in hashed_leaves that are in selected_indexes
+        bytes32[] memory indexed_leaves = new bytes32[](selected_indexes.length);
+        for (uint256 i = 0; i < selected_indexes.length; i++) {
+            indexed_leaves[i] = hashed_leaves[selected_indexes[i]];
+        }
+
+        bytes32 root = verifyComputeRoot(indexed_leaves, proof_hashes, proof_source_flags);
+
+        // Check if computed root is the same as the root of the tree
+        require(root == layers[layers.length - 1][0], "Invalid root");
+
+        // Convert proof_source_flags to bits and uint256
+        uint256 proof_flag_bits = 0;
+        bool[] memory proof_flag_bits_bool = new bool[](proof_source_flags.length);
+        for (uint256 i = 0; i < proof_source_flags.length; i++) {
+            if (proof_source_flags[i] == SOURCE_FROM_HASHES) {
+                proof_flag_bits_bool[i] = true;
+                proof_flag_bits = proof_flag_bits | (1 << i);
+            } else {
+                proof_flag_bits_bool[i] = false;
+                proof_flag_bits = proof_flag_bits | (0 << i);
+            }
+        }
+
+        return (proof_hashes, proof_flag_bits_bool, root);
+    }
+
+    /**
+     * @notice Generates a Merkle proof for a single leaf in the Merkle tree.
+     * @dev The function computes the proof and the root of the Merkle tree.
+     * @param leaves The array of leaves used to build the Merkle tree.
+     * @param leafIndex The index of the leaf for which the proof is generated.
+     * @return proof An array of sibling hashes forming the Merkle proof for the leaf.
+     * @return root The root hash of the Merkle tree.
+     */
+    function generateSingleProof(bytes32[] memory leaves, uint256 leafIndex)
+        public
+        pure
+        returns (bytes32[] memory, bytes32)
+    {
+        require(leaves.length > 1, "MerkleGen: Leaves should be greater than 1.");
+
+        // Append with the same leaf if odd number of leaves
+        if (leaves.length % 2 == 1) {
+            leaves = leaves.append(leaves[leaves.length - 1]);
+        }
+
+        bytes32[] memory proof = getProof(leaves, leafIndex);
+        bytes32 root = getRoot(leaves);
+
+        return (proof, root);
+    }
+
+    /**
+     * @notice Hashes two leaf nodes to generate their parent node.
      * @param a First child node.
      * @param b Second child node.
      * @return h Hashed parent node.
      */
-    function hash_internal_nodes(bytes32 a, bytes32 b) internal pure returns (bytes32 h) {
+    function hashLeafPairs(bytes32 a, bytes32 b) internal pure returns (bytes32 h) {
```

Comment:
Can we mark internal functions with prefix "_" to separate it from public functions?